adr/ADR-15.md

```diff
@@ -0,0 +1,235 @@
+
+For instance:
+- If a queue name is provided, configuring heartbeat or flow control is a mistake, since the server would send those message to random members.
+- If no subject is provided, along with no stream name, the library won't be able to locate/create the consumer.
+- For pull subscriptions ack policy of "none" or "all" is an error.
+
+At the time of this writing, the consumer configuration object looks like this (ordred alphabetically, not as seen in js.go):
+
+```go
+type ConsumerConfig struct {
+	AckPolicy       AckPolicy     `json:"ack_policy"`
+	AckWait         time.Duration `json:"ack_wait,omitempty"`
+	DeliverGroup    string        `json:"deliver_group,omitempty"`
+	DeliverPolicy   DeliverPolicy `json:"deliver_policy"`
+	DeliverSubject  string        `json:"deliver_subject,omitempty"`
+	Description     string        `json:"description,omitempty"`
+	Durable         string        `json:"durable_name,omitempty"`
+	FilterSubject   string        `json:"filter_subject,omitempty"`
+	FlowControl     bool          `json:"flow_control,omitempty"`
+	Heartbeat       time.Duration `json:"idle_heartbeat,omitempty"`
+	MaxAckPending   int           `json:"max_ack_pending,omitempty"`
+	MaxDeliver      int           `json:"max_deliver,omitempty"`
+	MaxWaiting      int           `json:"max_waiting,omitempty"`	// server returns error if provided if a DeliverSubject is provided
+	OptStartSeq     uint64        `json:"opt_start_seq,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_sequence
+	OptStartTime    *time.Time    `json:"opt_start_time,omitempty"`	// server returns error if provided when Deliver Policy is not by_start_time
+	RateLimit       uint64        `json:"rate_limit_bps,omitempty"` // Bits per sec
+	ReplayPolicy    ReplayPolicy  `json:"replay_policy"`
+	SampleFrequency string        `json:"sample_freq,omitempty"`
+}
+```
+
+#### Stream name
+
+When no stream is specified, the `subject` to the subscribe API calls is required. An error will be returned if it is not the case.
+The library will use the subject provided as a way to find out which stream this subscription is for. A request is sent to the server on the `<JS prefix>.STREAM.NAMES` subject with a JSON content `{"subject":"<subject here>"}`. If the response (a list of stream names) is positive and contains a single entry, then the library will use this stream name, otherwise an error indicating that there is no matching stream name should be returned.
+
+If the stream name is specified, then the `subject` is possibly not needed.
+
+#### Consumer name
+
+When a consumer name is specified, it indicates the the library that the intent is to bind to an existing JS Consumer. The library will lookup the consumer from the server and get a `ConsumerInfo`. If the lookup fails, it is considered an error, unless the subscription is for a "pull subscriber", in which case the library still proceeds with the JS subscription.
```

Comment:
1. can you change the word 'bind' to something like 'use' or 're-use'
2. Also can you use the word durable in this context with 'consumer name'

Reply:
1. ok
2. Sorry but no. You can reuse a consumer that is non-durable as long as you know its name. Sure it won't likely be the default behavior, but assuming that a consumer name means durable is misleading.